### ISSUE TITLE
Redirect conflict with uBlock Origin

### ISSUE THREAD

**Issue by terrorist96** [EFForg/https-everywhere#14961]
Type: other

```
This extension failed to redirect a network request to https://s7.addthis.com/js/300/addthis_widget.js#pubid=tenthamendment because another extension (uBlock Origin) redirected it to chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/web_accessible_resources/af3c919ee9a3d97f5dedaf2f8bec146a.javascript?secret=qqesz4qhqw5h.
```

Looks like a surrogate redirect conflict?

@gorhill

**Comment by blackhelicopters** [comment]
I'm getting the same 


`This extension failed to redirect a network request to https://www.google-analytics.com/analytics.js because another extension (uBlock Origin) redirected it to chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/web_accessible_resources/59fb24a2d12455d15bea20980e8a6801.javascript?secret=jblm6nqcej3g.`

**Comment by ariksc** [comment]
This extension failed to redirect a network request to https://s7.addthis.com/js/300/addthis_widget.js#pubid=ra-59c545b43f9cd35b because another extension (uBlock Origin) redirected it to chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/web_accessible_resources/af3c919ee9a3d97f5dedaf2f8bec146a.javascript?secret=nrw0ygwf8axk.

**Comment by gorhill** [comment]
https://github.com/gorhill/uBlock/issues/1170#issuecomment-168989068

**Comment by jeremyn** [comment]
@gorhill Your comment says "The resources used by uBO for redirection purpose are found [in this file](https://github.com/gorhill/uBlock/blob/master/assets/ublock/resources.txt)." but the linked file, https://github.com/gorhill/uBlock/blob/master/assets/ublock/resources.txt , doesn't exist anymore. Is there an equivalent?

**Comment by gorhill** [comment]
I corrected the link, the resource files have been moved to a different repo since I wrote this comment.

**Comment by mystica555** [comment]
I would appreciate this 'resources.txt' as exeptions, because honestly I don't want https everywhere to be redirecting me to tracking sites I explicitly block with ublock; the way it works now, trackers totally being blocked from loading, is the preferred scenario.  I just want the ! to disappear making me think there are 'actual' problems with  https on the site I am visiting.

**Comment by jeremyn** [comment]
@terrorist96 @blackhelicopters @ariksc @mystica555 Can you all please edit your comments to provide a URL where you go to see the problem? Not the error from the developer console, but the URL in your browser's address bar.

**Comment by terrorist96** [comment]
I'll try to post it if I come across it again. BTW, here's the commit Privacy Badger made to avoid this issue from occurring:
https://github.com/EFForg/privacybadger/pull/1905

**Comment by gorhill** [comment]
@jeremyn 

> we might take resources.txt and use it to create exceptions for the various rewrites it does

In that case `resources.txt` is not the right place for this, that would rather be the filters themselves, those with the `redirect=` option in any of the filter lists in <https://github.com/uBlockOrigin/uAssets/tree/master/filters>. There is a lots of these, and these lists are constantly edited. For the specific case here, this would be [this filter](https://github.com/uBlockOrigin/uAssets/blob/master/filters/privacy.txt#L82):

    ||addthis.com/*/addthis_widget.js$script,redirect=addthis.com/addthis_widget.js

**Comment by madusmacus** [comment]
This extension failed to redirect a network request to https://www.google-analytics.com/ga.js because another extension (uBlock Origin) 

I have also had 

This extension failed to redirect a network request to https://s7.addthis.com/js/300/addthis_widget.js#pubid=tenthamendment because another extension (uBlock Origin) redirected it to chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/web_accessible_resources

**Comment by jeremyn** [comment]
@gorhill Thanks. So it sounds like the logic for us might be: go through the files in https://github.com/uBlockOrigin/uAssets/tree/master/filters , look for filter lines that start with `||` and include `redirect=`, and add those partial-URLs to a list controlled by the checkbox I suggested earlier at https://github.com/EFForg/https-everywhere/issues/14961#issuecomment-375353304.

Why is issue popular suddenly? Did something change recently in uBlock Origin?

**Comment by gorhill** [comment]
> Why is this issue popular suddenly? Did something change recently in uBlock Origin?

I wondered. uBO has supported redirection since quite a while now, to `data:` URIs. However in a [recent release](https://github.com/gorhill/uBlock/releases/tag/1.15.12) uBO will redirect to web accessible resources instead of `data:` URIs. This has changed to solve long standing <https://github.com/gorhill/uBlock/issues/2823>. Not sure whether this is what is causing more reports about this -- redirecting to `data:` URIs was also causing such conflict before (as pointed out at <https://github.com/gorhill/uBlock/issues/1170>).

**Comment by jeremyn** [comment]
All of the reports here so far are reporting redirects to something under `web_accessible_resources`, so the recent uBlock Origin change may be related. But it sounds like these reporters should have seen the same error before, except it would have been to `data:` instead of `web_accessible_resources`, right?

**Comment by gorhill** [comment]
> it would have been to `data:` instead of `web_accessible_resources`, right?

Yes, it's what I find puzzling.

**Comment by Jed-Tech** [comment]
Been getting this on two different computers this week. Both with this message: 
`This extension failed to redirect a network request to https://widgets.outbrain.com/outbrain.js because another extension (uBlock Origin) redirected it to `chrome-extension:[LONGSTRING]/web_accessible_resources`

**Comment by jeremyn** [comment]
Pinging @Bisaloo @brainwane @gloomy-ghost @Hainish @J0WI @wonderchook to let them know about this.

**Comment by HvyMtlChaos** [comment]
I'm also getting this error in my extensions, the variant that cites google analytics and uBlockOrigin. I don't want to unblock that url in UBO and tried to find a way to exclude it from HTTPSEverywhere and couldn't.

**Comment by Bisaloo** [comment]
Since this is only a warning and does not affect the functionality of either uBlock or HTTPS Everywhere, I was going to vote for wontfix. But after having a look at the comments on Chrome store, it looks like it's making many users worried.

I'm still not sure how to fix this though...

**Comment by languitar** [comment]
There's a very obvious indication by chrome that something is wrong, which I have never seen before. People have to be worried with that icon appearing.

**Comment by Bisaloo** [comment]
> But it sounds like these reporters should have seen the same error before, except it would have been to data: instead of web_accessible_resources, right?

I did some digging on the Chromium bug tracker and this is why there wasn't any warnings before:

https://codereview.chromium.org/9857009/diff/1003/chrome/browser/extensions/api/web_request/web_request_api_helpers.cc

**Comment by cschanaj** [comment]
Possibly related https://bugs.chromium.org/p/chromium/issues/detail?id=800773

**Comment by gorhill** [comment]
> does not affect the functionality of either uBlock or HTTPS Everywhere

It could affect uBO: if the request is redirect by HTTP-Everywhere first, this will prevent uBO from redirecting to a local neutered resource, meaning a network request meant to be blocked would no longer be blocked. On the other hand, if uBO redirects first, this won't affect the purpose of HTTPS-Everywhere.

**Comment by cucumbur** [comment]
It affects the functionality when you don't use a myopic technical lens. HTTPSeverywhere and uBlock origin are fighting - reddit nerds ocndescend about using some matrix plugin and various other hardcore solutions that aren't practical for average users. They stop using either one and leave bad reviews - likely on HTTPSEverywhere, which is where I had to go see the big red error after chrome made it seem like my extensions were exploding. Less people WILL use the extension as a result.

**Comment by gorhill** [comment]
(Copied from https://github.com/gorhill/uBlock/issues/3639#issuecomment-376125053)

I won't tell people what to do. I will tell what is happening, and from this people can decide for themselves. One of two cases can happen:

The error message is a result of two extensions trying to redirect the same network request: the Chrome/WebExtensions API allows only one extension to redirect a network request. Extensions are not made aware that their redirection failed. The reason for this is probably technical: conceptually extensions execute in parallel.

### uBO redirects first

Benign.

Typically the error message _"This extension failed to redirect a network request..."_ will be with HTTPS-Everywhere in the Extensions page (as reported in [opening comment](https://github.com/EFForg/https-everywhere/issues/14961)).

This won't affect the purpose of HTTPS-Everywhere -- because no network requests is fired to the outside world.

### HTTPS-Everywhere redirects first

Serious.

Typically the error message _"This extension failed to redirect a network request..."_ will be with uBlock Origin in the Extensions page.

This will prevent uBO from redirecting the network request to a local neutered resource, meaning a network request which was meant meant to be blocked by uBO would no longer be blocked.

### Considerations

The order in which the extensions were installed is what decide which extension will redirect first (I can't remember the order) -- so this means by re-installing one or the other you could more from the "serious" to the "benign" case.

When a secure page (`https`) tries to pull a script from an unsecure location (`http`), the browser will refuse, mixed active content has been forbidden by browsers since quite a while now. So in that case, HTTPS-Everywhere is not really needed.

Why is this error more widely reported now? I have no clue, uBO has been redirecting to neutered resources since a very long time now (see #1170). Somebody on HTTPS-Everywhere's issue tracker [pointed out](https://github.com/EFForg/https-everywhere/issues/14961#issuecomment-376106816) that something related to extension error was fixed in Chrome 65.

**Comment by cschanaj** [comment]
AFIAK, using both uBO and HTTPSE together has been causing issues for a long time due to the limitation in the webRequest APIs (namely only one extension is allowed to rewrite/ block a request, without notifying the extension losing the race). This is becoming an critical issue when Chrome 65 start displaying the runtime errors in the UI. 

cc @Hainish as this might discourage the users from using HTTPSE...

**Comment by gorhill** [comment]
> only one extension is allowed to rewrite/ block a request

Blocking is not an issue -- if one extension orders a network request to be blocked, it will be blocked regardless of what other extensions decide. The issue is only when two or more extensions redirect. This is not an issue when the two extensions have similar purpose (uBO/Ghostery), but this here we have two extensions with different purpose.

**Comment by webvictim** [comment]
chrome://extensions

**Comment by jeremyn** [comment]
@webvictim Thanks, but I mean more like "If I go to https://example.com I get an error."

But it sounds like people are primarily seeing this reported against the extension, and not against the site, so that may be why people are having a difficult time identifying a specific site that triggers the error.

**Comment by gorhill** [comment]
I found a case, which might be the one from OP (I used the `pubid=tenthamendment` in his link to find a site with such URL): <http://blog.tenthamendmentcenter.com/>.

After testing teh case on my side, I realize I made a mistake is my assessment above: I did not take into account that the redirected network request by HTTPS will be further re-evaluated by all extensions, including uBO of course. This means that uBO will still properly redirect to its neutered resource after HTTPS-E redirects to the `https` version of the resource. Once the `https` version is fired, HTTPS-E will leave this one alone, while uBO will this time be able to redirect to its neutered local resources. I will correct my post above.

**Comment by jeremyn** [comment]
@gorhill Thanks, I can confirm that I get red error text at chrome://extensions/ for HTTPS Everywhere after I visit http://blog.tenthamendmentcenter.com/ while having both HTTPS Everywhere and uBlock Origin installed. I also originally got an exclamation mark on a yellow background near the top right of the Chrome window, which I somehow dismissed, and now I don't know how to get the exclamation mark back again. Also I'm not sure how to get rid of the red error text at chrome://extensions/ besides disabling and re-enabling HTTPS Everywhere.

**Comment by wonderchook** [comment]
I can confirm I am getting the same error. Would the user only notice it by going to chrome://extensions/ ?

**Comment by jeremyn** [comment]
@wonderchook The exclamation mark with the yellow background I mentioned was visible on the main screen.

**Comment by wonderchook** [comment]
@jeremyn okay, I see. I had recently updated Chrome so it is covered up by a green arrow asking me if I want to customize chrome. Once I dismissed the exclamation mark underneath the green arrow it appears not possible to get it back.

**Comment by NginUS** [comment]
This just started happening for me a few days ago. On 2 URLs that I've noticed. Got a screenshot of one, don't recall the other.
https://imgur.com/a/jqIds

**Comment by terrorist96** [comment]
Yes, the website was tenthamendmentcenter.com. Thanks @gorhill

**Comment by Hainish** [comment]
We could avoid this conflict by using the WebExtensions `onMessageExternal` handler: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/runtime/onMessageExternal.  If uBlock Origin sends a message to us saying "uBlock Origin is installed" or "uBlock origin is about to redirect so and so resource", then we could pre-emptively block our own redirects for these specific resources.

@gorhill can we collaborate on making this possible?

**Comment by Hainish** [comment]
@ghostwords has raised the point that this probably won't work, since `onMessageExternal` is an async handler.  The message will arrive too late, and blocking on message receipt would introduce an unacceptable delay.  This seems like something they'll have to fix on the Chromium side. There is a bug open for this: https://bugs.chromium.org/p/chromium/issues/detail?id=805704.  This issue does not resolve the conflicts between extensions, but seems like a good place to raise the point that scaring users with the current UX likely injures user retention rates.

**Comment by gorhill** [comment]
I think @Bisaloo's observation above is what should be looked into by Chromium devs: when an extension redirect to an external resource while another extension redirect to a local one, the redirection to the local resource should be seen as a blocked network request and no conflict warning should be raised -- as is currently done with `data:` URI,[1] and also for the same reason there is no conflict raised when an extension redirect while another block (block always win).

***

[1] So this explains why the conflict surfaced lately, it was silently ignored when uBO was relying on `data:` URI.

**Comment by selvodka** [comment]
me to having this issue mine is https://pastebin.com/raw/J4aKNF8H

**Comment by encip** [comment]
I get this issue on http://onemileatatime.boardingarea.com

**Comment by alundil** [comment]
This is the error I'm getting related to this issue as well.
https://i.imgur.com/EofRHzR.png

**Comment by J0WI** [comment]
IMHO this is a spec issue. Browsers or extensions need to handle the cases, where multiple extensions try to access the same API/resource at the same time.

**Comment by bartovan** [comment]
I now have the inverse error: uBlock Origin complains that another extension (HTTPS Everywhere) redirected first: 
"uBlock Origin
This extension failed to redirect a network request to chrome-extension://cjpalhdlnbpafiamejdnhcphjbkeiagm/web_accessible_resources/f6f25816c1c5ce3d465f4cd083539931.javascript?secret=gp6gzphd2wtw because another extension (HTTPS Everywhere) redirected it to https://widgets.outbrain.com/outbrain.js."

**Comment by Stellarspace** [comment]
I have been having this issue for weeks. I keep seeing it after I finish browsing.

**Comment by jeremyn** [comment]
Everyone: there is no need at this point to add a new comment simply to say you are getting the problem or to report the now-familiar `This extension failed to redirect a network request to chrome-extension://...`. Please instead thumbs-up the original comment if you are also getting the problem or want to show support for this issue.

**Comment by Jimbazuka** [comment]
I just want to point out the exact opposite happens to me where HTTPS everywhere succeeds in redirecting browser traffic over uBlockOrigin - what I see is documented in [here](https://www.reddit.com/r/uBlockOrigin/comments/87r1ax/ublock_origin_and_https_everywhere_extensions/)

**Comment by jeremyn** [comment]
To add to my previous comment: no one needs to add a comment to report an error happening in either direction, that is, `This extension failed to redirect a network request to https://...` or `This extension failed to redirect a network request to chrome-extension://...`. Please see @gorhill 's comment above at https://github.com/EFForg/https-everywhere/issues/14961#issuecomment-376127444 for a description of the problem.

I don't want to discourage people from showing their support for this issue, but rather, there is no need at the moment for anyone else to spend their time providing detailed information

**Comment by gorhill** [comment]
Consider starring this Chromium issue if you want a fix, I added a comment at the end:
https://bugs.chromium.org/p/chromium/issues/detail?id=111700#c31

**Comment by cowlicks** [comment]
I think it is worth noting the uBO change that is causing this, is actually a workaround for [this chrome bug](https://bugs.chromium.org/p/chromium/issues/detail?id=749236#c5). Which has implications for other projects (like PB). So please also chime in there to get this fixed.

**Comment by Hainish** [comment]
@gorhill as a temporary resolution to this problem, can you let HTTPS Everywhere know that uBlock Origin is installed via [onMessageExternal](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/runtime/onMessageExternal), and I will selectively disable the conflicting rulesets.  We would have to determine in advance which URLs are potentially problematic, but this is known, correct?  Meaning, it isn't dynamic, it's a static set of widget URLs which are locally proxied?

**Comment by gorhill** [comment]
@Hainish This is not as simple as just providing a list of URLs. The redirection is decided based on two things:

- Whether the network request is meant to be blocked
- Whether the network request matches a pattern in one of the directive in the redirection database

For the first one, this is decided at webRequest.onBeforeRequest listener time. The exact same network request may be blocked most of the time, but then allowed on a handful of sites (those sites broken by the act of blocking). It's decided in real-time, not in a declarative way.

The second point is that a redirection is decided according to a pattern in the URL, so a list of URLs does not work here. For example, a redirection directive with pattern `*/b/ss/*&aqe=`, which is meant to redirect to a local pixel image.

**Comment by jeremyn** [comment]
@Hainish If @gorhill doesn't have statistical data, then at least based on the comments here and https://github.com/gorhill/uBlock/issues/3639 , specific handling for Google Analytics would take care of a lot of cases.

**Comment by Paradoxum** [comment]
Is there any way to stop the (!) icon appearing at the very least for us end users? I'm getting this constantly and it's so frustrating.

**Comment by gorhill** [comment]
I added a [partial revert](https://github.com/gorhill/uBlock/commit/10a4a6465e213e7aa210bb0d0d8b6c7f7c8c7f89) of the fix to <https://github.com/gorhill/uBlock/issues/2823> (which is what the redirection to "web accessible resources" was fixing. uBO will avoid to redirect to a "web accessible resource" and instead redirect to a `data:` URI when the two following conditions are present:

- uBO is running in a Chromium-based browsers
- The URL to redirect is a non-secure `http:` (i.e. one that HTTP-Everywhere will likely want to redirect)

**Comment by lightheaded** [comment]
Just wanted to leave a note here that I'm still occasionally receiving this error on HTTPS Everywhere 2018.4.11

**Comment by Hainish** [comment]
@lightheaded thanks, I've pinged Chromium again on this issue: https://bugs.chromium.org/p/chromium/issues/detail?id=111700#c37

**Comment by gorhill** [comment]
@lightheaded Having the exact web page where this happens would be useful to determine whether this is really the same exact issue identified here.

**Comment by lightheaded** [comment]
@gorhill it's hard to track as it seemingly happens at random. But today I caught it blocking something on Amazon:

`This extension failed to redirect a network request to https://g-ecx.images-amazon.com/images/G/02/kindle/merch/2017/ACC/Promo/ACCBundles//UK_Bundle_Muscat_160X97.jpg because another extension (HTTPS Everywhere) redirected it to https://d1ge0kk1l5kms0.cloudfront.net/images/G/02/kindle/merch/2017/ACC/Promo/ACCBundles//UK_Bundle_Muscat_160X97.jpg`

**Comment by gorhill** [comment]
@lightheaded I am not aware of any filters in uBO which would want to redirect the URL your are mentioning.

Also, I find it surprising that HTTPS Everywhere would want to redirect a `https:` URL to another `https:` URL. The two related rules in HTTPS-E are to redirect from `http:` to `https:`:

- https://github.com/EFForg/https-everywhere/blob/73c41f6157287a02c42406e1b5d8115c7595f7a2/src/chrome/content/rules/Images-Amazon.com.xml#L31-L32
- https://github.com/EFForg/https-everywhere/blob/73c41f6157287a02c42406e1b5d8115c7595f7a2/src/chrome/content/rules/Amazon.xml#L116-L117

In any case, assuming you created custom rules to cause the behavior you are observing on your side, as said above, uBO fix with regard to HTTPS-E is a mitigation, i.e. this should take care of most conflicts, but this won't take care of conflict arising as a result of HTTPS-E redirecting a `https:` URL to another `https` URL (which I believe shouldn't be happening normally?)

**Comment by jeremyn** [comment]
@gorhill FYI, there's an open issue https://github.com/EFForg/https-everywhere/issues/10843 to remove `https:`->`https:` redirects where possible.

**Comment by cschanaj** [comment]
It seems that chrome will stop notifying the users for extension conflicts in the dev channel.

> This CL introduces the onActionIgnored event on the web request API which is
> fired when an extension's proposed modification to a network request is ignored.
> This happens in case of conflicts with other extensions. This should allow the
> extensions to address conflicts as they like.
> 
> Also, we no longer notify the user in case of conflicts between extensions.

https://groups.google.com/a/chromium.org/forum/#!msg/extensions-reviews/YYn1etdxjgw/55bBFxxfCgAJ

Ref

https://developer.chrome.com/extensions/webRequest#event-onActionIgnored

**Comment by Hainish** [comment]
Fantastic!  Thanks for the heads up @cschanaj - we should definitely handle this event.

**Comment by terrorist96** [comment]
Can this be closed? Marked as fixed since Aug 24.

> Consider starring this Chromium issue if you want a fix, I added a comment at the end:
> https://bugs.chromium.org/p/chromium/issues/detail?id=111700#c31

**Comment by Libbum** [comment]
Yep, nothing on my systems for a while too.
